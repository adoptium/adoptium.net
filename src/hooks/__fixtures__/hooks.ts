import {
  LatestTemurin,
  MockTemurinFeatureReleaseAPI,
  MockTemurinReleaseAPI,
  ReleaseNoteAPIResponse,
  ReleaseAsset,
  TemurinReleases,
} from "@/hooks";

export const createRandomTemurinRelease = (
  installer: boolean
): ReleaseAsset => ({
  os: "os_mock",
  architecture: "arch_mock",
  release_link: "https://release_link_mock",
  platform_name: "platform_name_mock",
  release_name: "release_name_mock",
  release_date: "2020-01-01T00:00:00Z",
  version: {
    major: 21,
    minor: 0,
    security: 1,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "21.0.1+0",
  },
  binaries: [
    {
      type: "type_mock",
      link: "https://link_mock",
      checksum: "checksum_mock",
      size: 0,
      extension: "extension_mock",
      release_name: "release_name_mock",
      installer_link: installer ? "https://installer_link_mock" : undefined,
      installer_checksum: installer ? "installer_checksum_mock" : undefined,
      installer_size: installer ? 0 : undefined,
      installer_extension: installer ? "installer_extension_mock" : undefined,
    },
  ],
});

export const mockLatestTemurin = (installer: boolean): LatestTemurin => ({
  download_count: 0,
  id: "id_mock",
  release_link: "https://release_link_mock",
  release_name: "release_name_mock",
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 21,
    minor: 0,
    security: 1,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "21.0.1+0",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "type_mock",
      heap_size: "heap_size_mock",
      download_count: 0,
      jvm_impl: "jvm_impl_mock",
      package: {
        name: "name_mock.tar.gz",
        link: "https://link_mock",
        checksum: "sha256sum_mock",
        checksum_link: "https://sha256sum_link_mock",
        metadata_link: "https://metadata_link_mock",
        signature_link: "https://signature_link_mock",
        size: 0,
        download_count: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            link: "https://installer_link_mock",
            checksum: "installer_sha256sum_mock",
            checksum_link: "https://installer_sha256sum_link_mock",
            metadata_link: "https://installer_metadata_link_mock",
            signature_link: "https://installer_signature_link_mock",
            size: 0,
            download_count: 0,
          }
        : undefined,
    },
  ],
});

export const createMockTemurinReleaseAPI = (
  installer: boolean,
  image_type: string
): MockTemurinReleaseAPI => ({
  release_link: "https://release_link_mock",
  release_name: "release_name_mock",
  vendor: "vendor_mock",
  version: {
    major: 21,
    minor: 0,
    security: 12,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "21.0.12+0",
  },
  binary: {
    architecture: "arch_mock",
    project: "project_mock",
    scm_ref: "scm_ref_mock",
    download_count: 0,
    updated_at: new Date(Date.UTC(2020, 0, 1)),
    os: "os_mock",
    image_type: image_type ? image_type : "jdk",
    jvm_impl: "jvm_impl_mock",
    heap_size: "heap_size_mock",
    package: {
      name: "name_mock.tar.gz",
      link: "https://link_mock",
      checksum: "checksum_mock",
      checksum_link: "https://checksum_link_mock",
      metadata_link: "https://metadata_link_mock",
      signature_link: "https://signature_link_mock",
      size: 0,
      download_count: 0,
    },
    installer: installer
      ? {
          name: "installer_name_mock.msi",
          link: "https://installer_link_mock",
          checksum: "installer_checksum_mock",
          checksum_link: "https://installer_checksum_link_mock",
          metadata_link: "https://installer_metadata_link_mock",
          signature_link: "https://installer_signature_link_mock",
          size: 0,
          download_count: 0,
        }
      : undefined,
  },
});

export const createMockReleaseNotesAPI = (
  number: number
): ReleaseNoteAPIResponse => ({
  id: "id_mock",
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "0.0.0+0",
  },
  release_name: "release_name_mock",
  // return an array of length number with the same release notes
  release_notes: Array.from({ length: number }, (v, i) => ({
    id: i.toString(),
    link: new URL("https://link_mock"),
    title: "title_mock",
    priority: "1",
    type: "Bug",
    component: "component_mock",
  })),
});

export const createMockTemurinFeatureReleaseAPI = (
  installer: boolean
): MockTemurinFeatureReleaseAPI => ({
  id: "id_mock",
  download_count: 0,
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "0.0.0+0",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "jdk",
      package: {
        name: "name_mock.tar.gz",
        checksum: "checksum_mock",
        link: new URL("https://link_mock"),
        size: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            checksum: "installer_checksum_mock",
            link: new URL("https://installer_link_mock"),
            size: 0,
          }
        : undefined,
    },
  ],
});

export const createRandomTemurinReleases = (
  installer: boolean
): TemurinReleases => ({
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  source_url: new URL("https://source_url_mock"),
  release_notes: true,
  release_notes_name: "release_notes_name_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  platforms: {
    platform_mock: {
      assets: [
        {
          os: "os_mock",
          architecture: "architecture_mock",
          type: "type_mock",
          link: new URL("https://link_mock"),
          checksum: "checksum_mock",
          size: 0,
          extension: "extension_mock",
          installer_link: installer
            ? new URL("https://installer_link_mock")
            : undefined,
          installer_checksum: installer ? "installer_checksum_mock" : undefined,
          installer_size: installer ? 0 : undefined,
          installer_extension: installer
            ? "installer_extension_mock"
            : undefined,
        },
      ],
    },
  },
});
