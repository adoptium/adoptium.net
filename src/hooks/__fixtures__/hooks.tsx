import {
  Binary,
  Contributor,
  ContributorApiResponse,
  EventAPI,
  LatestTemurin,
  MarketplaceRelease,
  MockTemurinFeatureReleaseAPI,
  MockTemurinReleaseAPI,
  ReleaseNoteAPIResponse,
  News,
  NewsResponse,
  ReleaseAsset,
  TemurinReleases,
  OperatingSystem,
  Architecture
} from "../hooks"

export const createRandomTemurinRelease = (installer): ReleaseAsset => ({
  os: "os_mock",
  architecture: "arch_mock",
  release_link: new URL("https://release_link_mock"),
  platform_name: "platform_name_mock",
  release_name: "release_name_mock",
  release_date: new Date(Date.UTC(2020, 0, 1)),
  binaries: [
    {
      type: "type_mock",
      link: new URL("https://link_mock"),
      checksum: "checksum_mock",
      size: 0,
      extension: "extension_mock",
      installer_link: installer
        ? new URL("https://installer_link_mock")
        : undefined,
      installer_checksum: installer ? "installer_checksum_mock" : undefined,
      installer_size: installer ? 0 : undefined,
      installer_extension: installer ? "installer_extension_mock" : undefined,
    },
  ],
})

export const createRandomMarketplaceRelease = (
  installer,
  id,
): MarketplaceRelease => ({
  release_name: "release_name_mock",
  vendor: "vendor_mock",
  binary: {
    os: "os_mock",
    architecture: "arch_mock",
    distribution: "distribution_mock",
    image_type: "type_mock",
    jvm_impl: "jvm_impl_mock",
    timestamp: new Date(Date.UTC(2020, 0, 1)),
    scm_ref: "scm_ref_mock",
    openjdk_scm_ref: "openjdk_scm_ref_mock",
    aqavit_results_link: new URL("https://aqavit_results_link_mock"),
    tck_affidavit_link: new URL("https://tck_affidavit_link_mock"),
    package: {
      name: "name_mock.tar.gz",
      link: new URL("https://link_mock"),
      sha256sum: `sha256sum_mock${id}`,
      sha256sum_link: new URL("https://sha256sum_link_mock"),
      signature_link: new URL("https://signature_link_mock"),
    },
    installer: installer
      ? [
          {
            name: "installer_name_mock.msi",
            link: new URL("https://installer_link_mock"),
            sha256sum: "installer_sha256sum_mock",
            sha256sum_link: new URL("https://installer_sha256sum_link_mock"),
            signature_link: new URL("https://installer_signature_link_mock"),
          },
        ]
      : undefined,
  },
  version: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
})

export const mockLatestTemurin = (installer): LatestTemurin => ({
  download_count: 0,
  id: "id_mock",
  release_link: new URL("https://release_link_mock"),
  release_name: "release_name_mock",
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 21,
    minor: 0,
    security: 1,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "type_mock",
      heap_size: "heap_size_mock",
      download_count: 0,
      jvm_impl: "jvm_impl_mock",
      package: {
        name: "name_mock.tar.gz",
        link: new URL("https://link_mock"),
        checksum: "sha256sum_mock",
        checksum_link: new URL("https://sha256sum_link_mock"),
        metadata_link: new URL("https://metadata_link_mock"),
        signature_link: new URL("https://signature_link_mock"),
        size: 0,
        download_count: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            link: new URL("https://installer_link_mock"),
            checksum: "installer_sha256sum_mock",
            checksum_link: new URL("https://installer_sha256sum_link_mock"),
            metadata_link: new URL("https://installer_metadata_link_mock"),
            signature_link: new URL("https://installer_signature_link_mock"),
            size: 0,
            download_count: 0,
          }
        : undefined,
    },
  ],
})

export const mockNewsAPI = (): NewsResponse => ({
  news: [
    {
      title: "news_title_mock",
      body: "body_mock",
      id: "id_mock",
      link: new URL("https://link_mock"),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ],
  pagination: {
    page: 1,
    pagesize: 5,
    result_start: 1,
    result_end: 5,
    result_size: 5,
    total_result_size: 16,
  },
})

export const mockEventsAPI = (): EventAPI => ({
  events: [
    {
      title: "events_title_mock",
      id: "id_mock",
      infoLink: new URL("https://link_mock"),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ],
})

export const createMockTemurinReleaseAPI = (
  installer,
  image_type,
): MockTemurinReleaseAPI => ({
  release_link: new URL("https://release_link_mock"),
  release_name: "release_name_mock",
  vendor: "vendor_mock",
  version: {
    major: 21,
    minor: 0,
    security: 12,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  binary: {
    architecture: "arch_mock",
    project: "project_mock",
    scm_ref: "scm_ref_mock",
    download_count: 0,
    updated_at: new Date(Date.UTC(2020, 0, 1)),
    os: "os_mock",
    image_type: image_type ? image_type : "jdk",
    jvm_impl: "jvm_impl_mock",
    heap_size: "heap_size_mock",
    package: {
      name: "name_mock.tar.gz",
      link: new URL("https://link_mock"),
      checksum: "checksum_mock",
      checksum_link: new URL("https://checksum_link_mock"),
      metadata_link: new URL("https://metadata_link_mock"),
      signature_link: new URL("https://signature_link_mock"),
      size: 0,
      download_count: 0,
    },
    installer: installer
      ? {
          name: "installer_name_mock.msi",
          link: new URL("https://installer_link_mock"),
          checksum: "installer_checksum_mock",
          checksum_link: new URL("https://installer_checksum_link_mock"),
          metadata_link: new URL("https://installer_metadata_link_mock"),
          signature_link: new URL("https://installer_signature_link_mock"),
          size: 0,
          download_count: 0,
        }
      : undefined,
  },
})

export const createMockReleaseNotesAPI = (number): ReleaseNoteAPIResponse => ({
  id: "id_mock",
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  release_name: "release_name_mock",
  // return an array of length number with the same release notes
  release_notes: Array.from({ length: number }, (v, i) => ({
    id: i.toString(),
    link: new URL("https://link_mock"),
    title: "title_mock",
    priority: "1",
    type: "Bug",
    component: "component_mock",
  })),
})

export const createMockTemurinFeatureReleaseAPI = (
  installer,
): MockTemurinFeatureReleaseAPI => ({
  id: "id_mock",
  download_count: 0,
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "jdk",
      package: {
        name: "name_mock.tar.gz",
        checksum: "checksum_mock",
        link: new URL("https://link_mock"),
        size: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            checksum: "installer_checksum_mock",
            link: new URL("https://installer_link_mock"),
            size: 0,
          }
        : undefined,
    },
  ],
})

export const createRandomTemurinReleases = (installer): TemurinReleases => ({
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  source_url: new URL("https://source_url_mock"),
  release_notes: true,
  release_notes_name: 'release_notes_name_mock',
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  platforms: {
    platform_mock: {
      assets: [
        {
          os: "os_mock",
          architecture: "architecture_mock",
          type: "type_mock",
          link: new URL("https://link_mock"),
          checksum: "checksum_mock",
          size: 0,
          extension: "extension_mock",
          installer_link: installer
            ? new URL("https://installer_link_mock")
            : undefined,
          installer_checksum: installer ? "installer_checksum_mock" : undefined,
          installer_size: installer ? 0 : undefined,
          installer_extension: installer
            ? "installer_extension_mock"
            : undefined,
        },
      ],
    },
  },
})

export const createRandomContributorViewData = (): Contributor => ({
  avatarUri: "avatarUri_mock",
  profileUri: "profileUri_mock",
  login: "login_mock",
  contributionsCount: 0,
  commitsListUri: "commitsListUri_mock",
  repo: "reponame_mock",
})

export const createRandomLatestForOSData = (): Binary => ({
  release_name: "release_name_mock",
  link: new URL("https://link_mock"),
})

export const createRandomNewsAndEventsData = (): News => ({
  news: {
    news: [
      {
        id: "id_mock",
        title: "title_mock",
        body: "body_mock",
        date: new Date(Date.UTC(2020, 0, 1)),
        link: new URL("https://link_mock"),
      },
    ],
    pager: {     
      current_page: 0,
      items_per_page: 5,
      total_items: 54,
      total_pages: 11,
    }
  },
  events: [
    {
      id: "id_mock",
      title: "title_mock",
      infoLink: new URL("https://infoLink_mock"),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ],
})

export const createRandomContributorViewData1Contribution =
  (): Contributor => ({
    avatarUri: "avatarUri_mock",
    profileUri: "profileUri_mock",
    login: "login_mock",
    contributionsCount: 1,
    commitsListUri: "commitsListUri_mock",
    repo: "reponame_mock",
  })

export const createMockAdoptiumContributorsApi =
  (): ContributorApiResponse => ({
    login: "test-user",
    id: 1,
    url: "https://api.github.com/users/test-user",
    type: "User",
    contributions: 10,
    node_id: "1",
    avatar_url: "https://github.com/images/error/octocat_happy.gif",
    gravatar_id: "",
    html_url: "https://github.com/images/error/octocat_happy.gif",
    followers_url: "https://api.github.com/users/test-user/followers",
    following_url:
      "https://api.github.com/users/test-user/following{/other_user}",
    gists_url: "https://api.github.com/users/test-user/gists{/gist_id}",
    starred_url:
      "https://api.github.com/users/test-user/starred{/owner}{/repo}",
    subscriptions_url: "https://api.github.com/users/test-user/subscriptions",
    organizations_url: "https://api.github.com/users/test-user/orgs",
    repos_url: "https://api.github.com/users/test-user/repos",
    events_url: "https://api.github.com/users/test-user/events{/privacy}",
    received_events_url:
      "https://api.github.com/users/test-user/received_events",
    site_admin: false,
  })

  export const mockOsesAPI = (): OperatingSystem[] => (
    [{name: "mock_macos", value: "mock_macos"}, {name: "mock_linux", value: "mock_linux"}, {name: "mock_windows", value: "mock_windows"}]
  )
  
  export const mockArchesAPI = (): Architecture[] => (
    [{name: "mock_aarch64", value: "mock_aarch64"}, {name: "mock_ppc64", value: "mock_ppc64"}, {name: "mock_x64", value: "mock_x64"}
  ])
  