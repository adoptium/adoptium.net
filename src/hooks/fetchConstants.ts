import { useEffect, useState } from "react";
import axios from "axios";
import { capitalize } from "@/utils/capitalize";

const baseUrl = "https://api.adoptium.net/v3";

export function useOses(isVisible: boolean): OperatingSystem[] {
  const [oses, setOses] = useState<OperatingSystem[]>([]);

  useEffect(() => {
    let isMounted = true;
    const cancelToken = axios.CancelToken.source();

    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/types/operating_systems`;

        axios
          .get(url, { cancelToken: cancelToken.token })
          .then(function (response) {
            if (!isMounted) return;

            const newOses = response.data
              .map((s: string) => {
                const o: OperatingSystem = {
                  name: capitalize(s),
                  value: s.toLowerCase(),
                };
                return o;
              })
              .sort((o1: OperatingSystem, o2: OperatingSystem) =>
                o1.name.localeCompare(o2.name)
              );

            setOses(newOses);
          })
          .catch(function (error) {
            if (!isMounted || axios.isCancel(error)) return;
            setOses([]);
          });
      })();
    }

    return () => {
      isMounted = false;
      cancelToken.cancel("Component unmounted");
    };
  }, [isVisible]);

  return oses;
}

export function useArches(isVisible: boolean): Architecture[] {
  const [arches, setArches] = useState<Architecture[]>([]);

  useEffect(() => {
    let isMounted = true;
    const cancelToken = axios.CancelToken.source();

    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/types/architectures`;

        axios
          .get(url, { cancelToken: cancelToken.token })
          .then(function (response) {
            if (!isMounted) return;

            const newArches = response.data
              .map((s: string) => {
                const a: Architecture = { name: s, value: s.toLowerCase() };
                if (a.name === "x32") a.name = "x86";
                return a;
              })
              .sort((a1: Architecture, a2: Architecture) =>
                a1.name.localeCompare(a2.name)
              );

            setArches(newArches);
          })
          .catch(function (error) {
            if (!isMounted || axios.isCancel(error)) return;
            setArches([]);
          });
      })();
    }

    return () => {
      isMounted = false;
      cancelToken.cancel("Component unmounted");
    };
  }, [isVisible]);

  return arches;
}

export interface OperatingSystem {
  name: string;
  value: string;
}

export interface Architecture {
  name: string;
  value: string;
}
