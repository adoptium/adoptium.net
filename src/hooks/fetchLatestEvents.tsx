import axios from "axios"

const baseUrl = "https://newsroom.eclipse.org/api"

export async function fetchLatestEvents() {
    const url = new URL(`${baseUrl}/events`)
    url.searchParams.append("parameters[publish_to]", "adoptium")
    url.searchParams.append("pagesize", "20")

    return axios
        .get(url.toString())
        .then(function (response) {
            return response.data.events
        })
        .catch(function () {
            return []
        })
}

export interface News {
    news: NewsResponse
    events: EventItem[]
}

export interface NewsResponse {
    news: NewsItem[]
    pager: {
        current_page: number
        items_per_page: number
        total_items: number
        total_pages: number
    } | null
}

export interface NewsItem {
    id: string
    title: string
    body: string
    date: Date
    link: URL
    image?: string  // Optional image URL from the Eclipse Foundation news API
}

export interface EventItem {
    id: string
    title: string
    infoLink: URL
    date: Date
}

export interface EventAPI {
    events: EventItem[]
}
