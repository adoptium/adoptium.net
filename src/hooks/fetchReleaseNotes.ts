import { VersionMetaData } from ".";
import { useEffect, useState } from "react";
import axios from "axios";

const baseUrl = "https://api.adoptium.net/v3/info/release_notes";
const releaseNamesUrl =
  "https://api.adoptium.net/v3/info/release_names?heap_size=normal&image_type=jdk&lts=true&page=0&page_size=1&project=jdk&release_type=ga&semver=false&sort_method=DEFAULT&sort_order=DESC&vendor=eclipse";

// Function to get the latest version string for a major version
async function getLatestVersionForMajor(
  majorVersion: string
): Promise<string | null> {
  try {
    // Use the assets API to get the latest release for this major version
    const url = `https://api.adoptium.net/v3/assets/latest/${majorVersion}/hotspot`;
    const response = await axios.get(url);

    if (
      response.data &&
      response.data.length > 0 &&
      response.data[0].release_name
    ) {
      return response.data[0].release_name;
    }
    return null;
  } catch (error) {
    console.error(
      `Failed to fetch latest version for major version ${majorVersion}:`,
      error
    );
    return null;
  }
}

export function useFetchReleaseNotesForVersion(
  isVisible: boolean,
  version: string | undefined,
  sortReleaseNotesByCallback?: (result: ReleaseNoteAPIResponse) => void
): ReleaseNoteDataBag | null {
  const [releaseNotes, setReleaseNotes] = useState<ReleaseNoteDataBag | null>(
    null
  );

  useEffect(() => {
    if (isVisible) {
      (async () => {
        let versionToDisplay = version;

        if (!versionToDisplay) {
          // retrieve the last version as default Release notes
          await axios
            .get(releaseNamesUrl)
            .then(function (response) {
              versionToDisplay = response.data.releases[0];
            })
            .catch(function () {
              // ignore
            });
        } else {
          // Check if the version is just a major version number (e.g., "8", "11", "17")
          const majorVersionRegex = /^\d+$/;
          if (majorVersionRegex.test(versionToDisplay)) {
            // Get the latest release for this major version
            const latestVersion = await getLatestVersionForMajor(
              versionToDisplay
            );
            if (latestVersion) {
              versionToDisplay = latestVersion;
            }
          }
        }

        const url = `${baseUrl}/${versionToDisplay}`;

        await axios
          .get(url.toString())
          .then(function (response) {
            const result = response.data;
            if (sortReleaseNotesByCallback) sortReleaseNotesByCallback(result);
            const releaseNoteDataBag: ReleaseNoteDataBag = {
              releaseNoteAPIResponse: result,
              isValid: result.release_notes !== null,
            };
            setReleaseNotes(releaseNoteDataBag);
          })
          .catch(function () {
            const releaseNoteDataBag: ReleaseNoteDataBag = {
              isValid: false,
            };
            setReleaseNotes(releaseNoteDataBag);
          });
      })();
    }
  }, [isVisible, version, sortReleaseNotesByCallback]);

  return releaseNotes;
}

export interface ReleaseNoteDataBag {
  releaseNoteAPIResponse?: ReleaseNoteAPIResponse;
  isValid: boolean;
}

export interface ReleaseNoteAPIResponse {
  id: string;
  release_name: string;
  release_notes: ReleaseNote[];
  vendor: string;
  version_data: VersionMetaData;
}

export interface ReleaseNote {
  id: string;
  link: URL;
  title: string;
  backportOf?: string;
  priority?: string;
  component?: string;
  subcomponent?: string;
  type?: string;
}
