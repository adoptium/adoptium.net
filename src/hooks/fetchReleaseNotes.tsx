import { VersionMetaData } from "."
import { useEffect, useState } from "react"
import axios from "axios"

const baseUrl = "https://api.adoptium.net/v3/info/release_notes"
const releaseNamesUrl = 'https://api.adoptium.net/v3/info/release_names?heap_size=normal&image_type=jdk&lts=true&page=0&page_size=1&project=jdk&release_type=ga&semver=false&sort_method=DEFAULT&sort_order=DESC&vendor=eclipse';

export function useFetchReleaseNotesForVersion(
  isVisible: boolean,
  version: string | undefined,
  sortReleaseNotesByCallback?: (result: ReleaseNoteAPIResponse) => void,
): ReleaseNoteDataBag | null {
  const [releaseNotes, setReleaseNotes] =
    useState<ReleaseNoteDataBag | null>(null)

  useEffect(() => {
    if (isVisible) {
      (async () => {
        let versionToDisplay = version;

        if (!versionToDisplay) {
          // retrieve the last version as default Release notes
          await axios.get(releaseNamesUrl)
            .then(function (response) {
              versionToDisplay = response.data.releases[0];
            })
            .catch(function () {
              // ignore
            });
        }

        const url = `${baseUrl}/${versionToDisplay}`;

        console.log(`Fetching release notes for version: ${versionToDisplay} from URL: ${url}`)

        await axios
          .get(url.toString())
          .then(function (response) {
            const result = response.data
            if (sortReleaseNotesByCallback) sortReleaseNotesByCallback(result)
            const releaseNoteDataBag: ReleaseNoteDataBag = {
              releaseNoteAPIResponse: result,
              isValid: (result.release_notes !== null)
            };
            setReleaseNotes(releaseNoteDataBag);
          })
          .catch(function () {
            const releaseNoteDataBag: ReleaseNoteDataBag = {
              isValid: false
            };
            setReleaseNotes(releaseNoteDataBag);
          })
      })()
    }
  }, [isVisible, version, sortReleaseNotesByCallback])

  return releaseNotes
}

export interface ReleaseNoteDataBag {
  releaseNoteAPIResponse?: ReleaseNoteAPIResponse
  isValid: boolean
}

export interface ReleaseNoteAPIResponse {
  id: string
  release_name: string
  release_notes: ReleaseNote[]
  vendor: string
  version_data: VersionMetaData
}

export interface ReleaseNote {
  id: string
  link: URL
  title: string
  backportOf?: string
  priority?: string
  component?: string
  subcomponent?: string
  type?: string
}
