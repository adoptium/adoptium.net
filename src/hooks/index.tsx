'use client'

export * from "./useOnScreen"
export * from "./fetchLatestTemurin"
export * from "./fetchTemurinReleases"
export * from "./fetchMarketplace"

export interface VersionMetaData {
    major: number
    minor: number
    security: number
    pre?: string
    patch?: number
    build: number
    optional?: string
    adopt_build_number?: number
    openjdk_version: string
    semver: string
}

/**
 * This method compares two versions and returns:
 * ->  0 if versions are equals
 * -> -1 if versionA is lower than versionB
 * ->  1 if versionA is highter than versionB
 *
 * @param versionA
 * @param versionB
 * @returns
 */
export function compareVersions(
    versionA: VersionMetaData,
    versionB: VersionMetaData,
): number {
    let comparison = versionA.major - versionB.major
    if (comparison === 0) {
        comparison = versionA.minor - versionB.minor
        if (comparison === 0) {
            comparison = versionA.security - versionB.security
            if (comparison === 0) {
                comparison = versionA.build - versionB.build
                if (comparison === 0) {
                    comparison =
                        (versionA.adopt_build_number ? versionA.adopt_build_number : 0) -
                        (versionB.adopt_build_number ? versionB.adopt_build_number : 0)
                }
            }
        }
    }
    return comparison
}

export function getVersionAsString(
    version: VersionMetaData,
    allowShortNotation: boolean = false,
): string {
    let result = `${version.major}.${version.minor}.${version.security}`
    if (
        allowShortNotation &&
        !version.minor &&
        !version.security &&
        !version.patch
    ) {
        // when minor, security & patch are 0 => do not print .0.0.0
        result = `${version.major}`
    }
    if (version.patch) {
        result += `.${version.patch}`
    }
    if (version.pre) {
        result += `-${version.pre}`
    }
    if (version.build) {
        result += `+${version.build}`
        if (version.adopt_build_number) result += `.${version.adopt_build_number}`
    }
    if (version.optional) {
        result += `-${version.optional}`
    }
    return result
}
