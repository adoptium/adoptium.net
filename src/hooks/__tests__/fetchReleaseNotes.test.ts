import { renderHook, waitFor } from "@testing-library/react";
import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { useFetchReleaseNotesForVersion } from "../fetchReleaseNotes";
import { createMockReleaseNotesAPI } from "@/hooks/__fixtures__/hooks";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

const mock = new MockAdapter(axios);
const mockResponse = createMockReleaseNotesAPI(1);

afterEach(() => {
  vi.clearAllMocks();
  mock.reset();
});

afterAll(() => {
  mock.restore();
});

const sortReleaseNotesByCallback = vi.fn();

describe("useFetchReleaseNotesForVersion", () => {
  it("returns valid JSON", async () => {
    mock.onGet().reply(200, mockResponse);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() =>
      useFetchReleaseNotesForVersion(
        true,
        "sample_version",
        sortReleaseNotesByCallback
      )
    );
    await waitFor(
      () => {
        expect(result.current?.releaseNoteAPIResponse?.release_name).toBe(
          "release_name_mock"
        );
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      "https://api.adoptium.net/v3/info/release_notes/sample_version"
    );
    expect(result.current).toMatchSnapshot();
  });

  it("returns null if error is caught in fetch", async () => {
    mock.onGet().reply(200, mockResponse);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() =>
      useFetchReleaseNotesForVersion(
        true,
        "sample_version",
        sortReleaseNotesByCallback
      )
    );
    await waitFor(
      () => {
        expect(result.current).toBe(null);
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      "https://api.adoptium.net/v3/info/release_notes/sample_version"
    );
  });

  it("returns null if version is not defined", async () => {
    mock.onGet().reply(200, mockResponse);

    const { result } = renderHook(() =>
      // @ts-expect-error version is intentionally undefined
      useFetchReleaseNotesForVersion(true, null, sortReleaseNotesByCallback)
    );
    expect(result.current).toBe(null);
  });

  it("checks sortReleaseNotesByCallback to be called", async () => {
    mock.onGet().reply(200, mockResponse);

    renderHook(() =>
      useFetchReleaseNotesForVersion(
        true,
        "sample_version",
        sortReleaseNotesByCallback
      )
    );
    await waitFor(
      () => {
        expect(sortReleaseNotesByCallback).toHaveBeenCalledTimes(1);
      },
      { interval: 1 }
    );
  });

  it("checks sortReleaseNotesByCallback NOT to be called", async () => {
    mock.onGet().reply(200, mockResponse);

    renderHook(() => useFetchReleaseNotesForVersion(true, "sample_version"));
    await waitFor(
      () => {
        expect(sortReleaseNotesByCallback).toHaveBeenCalledTimes(0);
      },
      { interval: 1 }
    );
  });

  it("releaseNotes to be null on error", async () => {
    mock.onGet().reply(500);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() =>
      useFetchReleaseNotesForVersion(true, "sample_version")
    );
    await waitFor(
      () => {
        expect(result.current).toEqual({ isValid: false });
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
