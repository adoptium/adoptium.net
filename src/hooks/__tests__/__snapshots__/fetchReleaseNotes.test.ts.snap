// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFetchReleaseNotesForVersion > returns valid JSON 1`] = `
{
  "isValid": true,
  "releaseNoteAPIResponse": {
    "id": "id_mock",
    "release_name": "release_name_mock",
    "release_notes": [
      {
        "component": "component_mock",
        "id": "0",
        "link": "https://link_mock/",
        "priority": "1",
        "title": "title_mock",
        "type": "Bug",
      },
    ],
    "vendor": "vendor_mock",
    "version_data": {
      "build": 0,
      "major": 0,
      "minor": 0,
      "openjdk_version": "openjdk_version_mock",
      "patch": 0,
      "security": 0,
      "semver": "0.0.0+0",
    },
  },
}
`;
